### PR TITLE
Migrate to the modern linker input API.

### PR DESCRIPTION
**Pull request by benjaminp**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

Migrates cgo code away from deprecated Bazel cc Starlark apis.

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

See https://github.com/bazelbuild/bazel/issues/10860.

### CODE REVIEW

**Comment by benjaminp** [comment]
I would also need to change the minimum Bazel version to 3.0.0.

File: go/private/common.bzl [hunk 1]
```diff
@@ -208,7 +208,7 @@ def has_versioned_shared_lib_extension(path):
 
     return True
 
-MINIMUM_BAZEL_VERSION = "2.2.0"
+MINIMUM_BAZEL_VERSION = "3.0.0"
```
Comment (jayconrod) [inline go/private/common.bzl:211]: Please also update the minimum version in README.rst.

**Review by jayconrod — COMMENTED** [review]
Change looks good. CI is down right now (bazelbuild/continuous-integration#1055), so I'll check back on this later.

File: go/private/common.bzl (hunk 1 above)
Comment (benjaminp) [reply go/private/common.bzl:211]: Okay, I wasn't sure if I should do that, since it specifically references the latest released version of rules_go.

Comment (jayconrod) [reply go/private/common.bzl:211]: You're right, it's actually better to leave that for the next minor release. Forget my earlier comment.

**Pull request merged**